packages/transaction-controller/src/history.ts

```diff
@@ -0,0 +1,86 @@
+import jsonDiffer from 'fast-json-patch';
+import { cloneDeep } from 'lodash';
+
+import type {
+  TransactionHistory,
+  TransactionHistoryEntry,
+  TransactionMeta,
+} from './types';
+
+/**
+ * Generates a history entry from the previous and new transaction metadata.
+ *
+ * @param previousState - The previous transaction metadata.
+ * @param currentState - The new transaction metadata.
+ * @param note - A note for the transaction metada update.
+ * @returns An array of history operation.
+ */
+function generateHistoryEntry(
+  previousState: any,
+  currentState: TransactionMeta,
+  note: string,
+): TransactionHistoryEntry {
+  const historyOperationsEntry = jsonDiffer.compare(
+    previousState,
+    currentState,
+  ) as TransactionHistoryEntry;
+  // Add a note to the first operation, since it breaks if we append it to the entry
+  if (historyOperationsEntry[0]) {
+    if (note) {
+      historyOperationsEntry[0].note = note;
+    }
+    historyOperationsEntry[0].timestamp = Date.now();
+  }
+  return historyOperationsEntry;
+}
+
+/**
+ * Compares and adds history entry to the provided transactionMeta history.
+ *
+ * @param transactionMeta - TransactionMeta to add history entry to.
+ * @param note - Note to add to history entry.
+ */
+export function updateTransactionHistory(
```

Comment:
Should we keep the exported functions at the top to highlight them?

Reply:
Fixed in https://github.com/MetaMask/core/pull/1657/commits/577203e91035b5bb13a5b2c7151f255abe70f582